doc/developer/design/20230119_ir_normalization.md

```diff
@@ -0,0 +1,147 @@
+
+WARNING: `Filter` operators do not introduce a specific evaluation order, and moving them is a recurring source of stress when it is used to guard erroring computation.
+We are "allowed" to commute it wherever we want, but we may need to push back on constructs that try to use `Filter` as an operator that "guarantees" much of anything, as it doesn't guarantee very much about the act of execution and evaluation, and once you panic (read: "introduce an record into the error output") you can't easily undo it.
+
+### Fusion
+
+Stacks of equivalent operators can often be consolidated into (at most) one operator.
+For example, each of `Map`, `Filter`, `Project`, `Union`, `Join`, `Negate`.
+Some forms of `Reduce` and `TopK` can be fused (`Distinct`, and same (key, order) respectively).
+We can restrict our attention to operators that can be fused without further investigation of the properties of their inputs.
+
+### Canonicalization
+
+Each operator should canonicalize its representation. 
+`Filter` should sort, deduplicate, and simplify its predicates.
+`Union` and `Join` should (could?) sort their inputs and permute their outputs.
+`Reduce` can sort its `group_key` and `aggregate` arguments.
+All operators should `reduce` their argument expressions to their simplest forms.
+
+Several of these proposals *introduce* permutations.
+These permutations are new AST nodes, but they should be easy to remove as part of a post-order traversal.
+
+### Elision
+
+Operators that no longer need to exist should be removed.
+For example,
+* Maps and Filters with empty arguments,
+* Projections that take no action,
+* Joins and Unions with single inputs (and no constraints),
+
+If we introduce permutations in prior steps, this is also a great moment to remove them.
+Similarly, we may wish to absorb all projections in to certain parent operators (e.g. `Reduce`).
+
+### Constant Folding
+
+Collections that are entirely constants can be simplified.
+This is not the same as constant value propagation for columns; let's not do that here.
+If you have a seven element collection and a filter on it, apply the filter
+
+
+## Alternatives
+
+<!--
+// Similar to the Description section. List of alternative approaches considered, pros/cons or why they were not chosen
+-->
+
+The alternative is to not attempt to normalize expressions, or to normalize them less.
+
+I think we will learn some about the reasons for normalization, and we may conclude that only some of the proposed normalizations have value. 
+
+## Open questions
+
+<!--
+// Anything currently unanswered that needs specific focus. This section may be expanded during the doc meeting as
+// other unknowns are pointed out.
+// These questions may be technical, product, or anything in-between.
+-->
+
+The specific normalizations that have value are very much up for discussion.
+The tests we use to determine if something belongs as a normalization, e.g. that it is only removes ambiguity, are up for discussion.
```

Comment:
Re-stating what I said above, my current take is that "something belongs as a normalization" is that it allows follow-up transformations to simplify their reasoning about the input patterns that they can observe.